[pytorch] Add native support for segment reduce step1: API definition

serhaty opened the pull request:
Summary:
This is first diff to add native support for segment reduction in PyTorch. It provides similar functionality like torch.scatter or "numpy.ufunc.reduceat".

This diff mainly focuses on API layer to make sure future improvements will not cause backward compatibility issues. Once API is settled, here are next steps I am planning:
- Add support for other major reduction types (e.g. min, sum) for 1D tensor
- Add Cuda support
- Backward support
- Documentation for the op
- Perf optimizations and benchmark util
- Support for multi dimensional tensors (on data and lengths) (not high priority)
- Support for 'indices' (not high priority)

Test Plan: Added unit test

Differential Revision: D26952075

ngimel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9289,3 +9289,8 @@
   cpp_no_default_args: ['a', 'b']
   python_module: nn
+
+- func: segment_reduce(Tensor data, str reduce, Tensor? lengths=None, Tensor? indices=None, int axis=0, bool unsafe=False) -> Tensor
```
you probably want all the args except `data` to be kwarg-only, so do it like
```
segment_reduce(Tensor data, *, str reduce ....)
```

serhaty replied:
Is there an example I can follow for this? Also, where would the default be set in this approach?

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+    const c10::optional<Tensor>& indices,
+    int64_t axis,
+    bool unsafe) {
+  TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
```
pytorch usually supports wrapping of dimensions, so e.g. 0 and -1 would be valid `axis` argument for a 1d tensor, so before checking `axis` do 
```
axis = maybe_wrap_dim(axis, data.ndim());
```

serhaty replied:
you meant "ndimension", right? SG! updated.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+    bool unsafe) {
+  TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
+  TORCH_CHECK(data.dim() == 1);
+  TORCH_CHECK(!data.is_cuda());
```
this check is not needed, dispatcher will take care of this

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+  TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
+  TORCH_CHECK(data.dim() == 1);
+  TORCH_CHECK(!data.is_cuda());
+  TORCH_CHECK(data.numel() > 0);
```
if possible, you should support empty tensors, by returning empty output tensor with the corresponding sizes.

serhaty replied:
sg, will follow-up in future diff with test as well.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+  if (!unsafe) {
+    int64_t sum = 0;
+    for (int64_t i = 0; i < batch_size; ++i) {
+      sum += lengths_data[i];
```
you should also be checking that length_data[i] is not negative here.

serhaty replied:
Also checking 0 as well now, to avoid issue below. So if unsafe is True, we will set to numerical limit based on reduction, otherwise check of valid lengths here. Let me know if this makes sense.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+    TORCH_CHECK(sum == data.numel());
+  }
+
+  AT_DISPATCH_ALL_TYPES(
```
ALL is not really ALL in pytorch, so if you want to support e.g. bfloat16 you'd need to add it separately with DISPATCH_ALL_TYPES_AND

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+        for (int64_t i = 0; i < batch_size; ++i) {
+          scalar_t reduction = std::numeric_limits<scalar_t>::lowest();
+          for (int64_t j = 0; j < lengths_data[i]; ++j) {
+            if (values_data[k] > reduction) {
```
this does not properly handle nan values, you need something like
reduction = at::_isnan(data) ? data : std::max<scalar_t>(reduction,data); //std::max returns first arg if one of the args is nan

serhaty replied:
Thanks for the pointer. Updated. I think you meant "at::_isnan(data) ? reduction ..." instead, right?

ngimel replied:
It should be `data`, not `reduction`, to properly propagate `nan` (also, will be good to add a test).

serhaty replied:
I actually updated the test in the PR with none value. I guess expected behaviour is not clear for me. If one of the values is None for a reduction, do we want to return None reduction or result based on remaining items within the segment?

I don't have preference one way or another but would like to keep similar convention as rest of PyTorch regarding none handling. If returning None is preferred, will update the test and PR.

ngimel replied:
Default is propagate nan values, that is, if nan is encountered in the segment, it is returned. But the expression should be as in my first comment to return nan if either of the args is nan
```
reduction = at::_isnan(data) ? data : std::max<scalar_t>(reduction,data); //std::max returns first arg if one of the args is nan
```
If `data` is `nan` it is returned, otherwise if `reduction` is `nan` it will be returned as the first arg to `std::max`, otherwise max(reduction, data) is returned.

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -0,0 +1,81 @@
+          if (lengths_data[i] > 0) {
+            output_data[i] = reduction;
+          } else {
+            output_data[i] = (scalar_t)0;
```
for max reduction, you can't reduce over 0-length segments, 0 is not a proper identity element for reduction. 0 length is an invalid input for a max reduction.

ngimel commented on test/test_segment_reductions.py:
```diff
@@ -0,0 +1,33 @@
+    def test_max_simple_1d(self, device, dtype):
+        lengths = torch.tensor([1, 2, 3], device=device)
+        data = torch.tensor([1, float("nan"), 3, 4, 5, 6], device=device, dtype=dtype)
+        expected_result = torch.tensor([1, 3, 6], device=device, dtype=dtype)
```
Typically, operations in pytorch propagate `nans` - do you need different behavior here?

serhaty replied:
replied in earlier comment. Will update the PR to return nan instead of reduction based on remaining elements.

ngimel reviewed (commented):
Looks good, there's an issue of desired nan propagation behavior.

xw285cornell commented on test/test_segment_reductions.py:
```diff
@@ -0,0 +1,37 @@
+    @onlyCPU
+    @dtypes(torch.half, torch.bfloat16, torch.float, torch.double)
+    def test_max_simple_1d(self, device, dtype):
+        lengths = torch.tensor([1, 2, 3], device=device)
```
usually it'll be a good idea to use hypothesis to test more cases w/ random number, and use numpy to write a ref implementation?

ngimel replied:
We typically don't use hypothesis in torch testsuite

xw285cornell replied:
I see, is it because of pulling extra dependencies? JUst try to think how to make the test stronger and cover corner cases; Maybe we can use torch.randint to randomize the data (btw in the past we do see some corner cases which fails the job, e.g. when input data is empty)

ngimel replied:
We usually specifically test for corner-case inputs instead of hoping that they'd be randomly generated

serhaty replied:
I still have todo to add more testing, documentation and benchmark in follow-up diffs. So this is still WIP.

Facebook:
Once we have high confidence on the op, only then will switch Dper3's LengthsMax and variations to segment_reduce.

serhaty replied:
I still have todo to add more testing, documentation and benchmark in follow-up diffs. So this is still WIP.

xw285cornell reviewed (commented):
Nice :)

ngimel commented:
This looks good, can you please rebase and reexport to get CI signal? There are 17 failing tests, and even though they are failing due to connection issues, we need to get signal from them.

Commit:
[pytorch] Add native support for segment reduce step1: API definition (#53727)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/53727

This is first diff to add native support for segment reduction in PyTorch. It provides similar functionality like torch.scatter or "numpy.ufunc.reduceat".

This diff mainly focuses on API layer to make sure future improvements will not cause backward compatibility issues. Once API is settled, here are next steps I am planning:
- Add support for other major reduction types (e.g. min, sum) for 1D tensor
- Add Cuda support
- Backward support
- Documentation for the op
- Perf optimizations and benchmark util
- Support for multi dimensional tensors (on data and lengths) (not high priority)
- Support for 'indices' (not high priority)

Test Plan: Added unit test

Differential Revision: D26952075

fbshipit-source-id: 3928f127b8d1e58b2a10be4c9ded320604799849
